### ISSUE TITLE
Mouse callback gets wrong coordinates with small images

### ISSUE THREAD

**Issue by pschroeppel** [opencv/opencv#7485]
##### System information (version)

- OpenCV => 3.1.0
- Operating System / Platform => Mac OSX El Capitan 10.11.6
- Compiler => Apple LLVM version 8.0.0 (clang-800.0.38)

##### Detailed description

I load an image with imread() and display it with imshow(). On the window I set a mouse callback that gives out the coordinates when the image is clicked with the left mouse button.
But when the image is too small (for example 100x100px) there is a gray "filling" area in the window where the image is displayed. The problem is, that this results in wrong coordinates when the image is clicked:
-The window seems to be 200x100px with 100x100px of the real image and 100x100px of gray filling area
-When I click on the right bottom of the image I get the coordinates (49,99) and when I click on the right bottom of the whole window I get the coordinates (99,99)

My teacher compiled the same program on Linux and the image is displayed correctly in a window of 100x100px and correct coordinates in the mouse callback.

##### Steps to reproduce

I attached a zip file which contains:
-source code
-input image with 100x100px which leads to the error (color_test.jpg)
-input image with 300x300px which works correctly
-screenshot of how the 100x100px image is displayed with the gray filling area (output.jpg)

[onmouse_bug.zip](https://github.com/opencv/opencv/files/527658/onmouse_bug.zip)

**Comment by StevenPuttemans** [comment]
To avoid this
1. Make sure you build OpenCV without Qt support. It makes these debug windows way larger then they should be to fit all options in...
2. Make sure your window name is small

Then problems will all go away!

**Comment by bedbad** [comment]
It is Apple bug in Cocoa in 
window setAutosize(bool)
There is no issue on all other systems. Including Qt support. 
It works just fine with Qt(qt5) support on Linux(Fedora, Debian..)

**Comment by sunsided** [comment]
It's even worse with the Qt-based GUI interfering with mouse and key events and setting the `WINDOW_GUI_NORMAL` flag doesn't change that. Shift-leftclick e.g. enables panning, but does not change the mouse coordinates accordingly.

**Comment by sjmgarnier** [comment]
@alalek :thumbsup:

**Comment by sjmgarnier** [comment]
@alalek I have compiled OpenCV using the patch, it does seem to speed things up a bit, but does not seem to solve the problem with the coordinates unfortunately.

**Comment by tellowkrinkle** [comment]
I updated the patch, does it work properly now?

**Comment by tellowkrinkle** [comment]
Alright I'll look into that

**Comment by tellowkrinkle** [comment]
I don't seem to be having the issue... I tested the code in `onmouse_bug.zip` and it displays an image immediately looking like this, which I assume is correct 
<img width="200" alt="image" src="https://user-images.githubusercontent.com/3315070/46448180-24486900-c74b-11e8-8be7-28d93f5b8357.png">

**Comment by sjmgarnier** [comment]
@tellowkrinkle Ok. Let me take a look at my code again. It's likely that I'm doing something wrong.

**Comment by sjmgarnier** [comment]
@tellowkrinkle Sorry I can't find onmouse_bug.zip.

**Comment by tellowkrinkle** [comment]
It's at the end of the very first post in this thread

**Comment by sjmgarnier** [comment]
And here I was scouring the entire opencv source code to find it :-)

**Comment by tellowkrinkle** [comment]
Yeah admittedly I didn't really notice that you weren't the person who had originally made this thread so my first thought was "wasn't that your own example?"

**Comment by sjmgarnier** [comment]
Not sure what's up. My code is essentially similar to the one in the example, but I get a gray image the first time I want to display an image in a new window. If I call imshow to an already existing window (like the one with the gray image), no more problem. I'm guessing the issue is on my end if you can't reproduce it, so don't worry about it.

**Comment by tellowkrinkle** [comment]
Does it work with the old window system, and is it something you're willing to post here?

**Comment by sjmgarnier** [comment]
It was, but I'll check again tomorrow to make sure. It's getting late here, logging off for the night.

**Comment by sjmgarnier** [comment]
@tellowkrinkle It works with the old system. My code is here: https://github.com/swarm-lab/Rvision/blob/master/src/display.h.

**Comment by tellowkrinkle** [comment]
Do you have a specific main file I can use?  I tried this with the `display.h` code pasted above it both with and without the call to `_newDisplay` and both worked as expected (displayed a very green window) 
```c++
int main(int argc, const char * argv[]) {
	cv::Mat test = cv::Mat(cv::Size(512, 512), CV_8UC3);
	test.setTo(cv::Scalar(0, 255, 128));
	_newDisplay("Test", 512, 512);
	_display(test, "Test", 0, 512, 512);
	return 0;
}
```

**Comment by sjmgarnier** [comment]
@tellowkrinkle I don't, this is part of a R package (two technically). On my end, calling _display doesn't show the image unless the window is already created. I don't think you should worry too much about it, the cause might be with the way I included your changes before compiling OpenCV. I simply replaced window_cocoa.mm in the 3.4.3 release version before compiling OpenCV (I'm guessing that you're using the latest version of OpenCV instead). I also made a few changes to OpenCVModule.3.4.3.cmake. I will need first to investigate whether these might be the cause of the problem. If it works on your end, then I suspect the problem is with my modifications rather than yours.

**Comment by tellowkrinkle** [comment]
Alright well if you do end up finding an issue make sure to mention me when you report it, otherwise I doubt I'll see it.
